src/ipc/ipc4/mtrace.c

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Kai Vehmanen <kai.vehmanen@linux.intel.com>
+
+#include <stdint.h>
+#include <sof/lib/uuid.h>
+#include <sof/ipc/common.h>
+#include <ipc4/base_fw.h>
+
+#undef panic // SOF headers may redefine panic
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_backend_adsp_mtrace.h>
+#include <zephyr/logging/log.h>
+
+#include <sof/schedule/ll_schedule.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/task.h>
+
+LOG_MODULE_REGISTER(mtrace, CONFIG_SOF_LOG_LEVEL);
+
+/**
+ * If mtrace log buffer has less space than
+ * the threshold, notify host with BUFFER_STATUS IPC4.
+ */
+#define NOTIFY_BUFFER_STATUS_THRESHOLD 2048
+
+/**
+ * Maximum time to block IPC BUFFER STATUS notify messages.
+ * The notification is sent either due to amount of data
+ * available in the buffer, or by timeout in case logs are
+ * sent at slow rate.
+ */
+#define IPC4_MTRACE_NOTIFY_MAX_DELAY_MS 1000
+
+#ifdef CONFIG_LOG_BACKEND_ADSP_MTRACE
+
+/* bb2aa22e-1ab6-4650-8501-6e67fcc04f4e */
+DECLARE_SOF_UUID("mtrace-task", mtrace_task_uuid, 0xbb2aa22e, 0x1ab6, 0x4650,
+		 0x85, 0x01, 0x6e, 0x67, 0xfc, 0xc0, 0x4f, 0x4e);
+
+static uint64_t msg_notify_last_sent;
+
+static uint32_t mtrace_pending;
+
+static uint32_t mtrace_aging_timer = IPC4_MTRACE_NOTIFY_MAX_DELAY_MS;
+
+struct task mtrace_work;
+
+static void mtrace_log_hook(size_t written, size_t space_left)
+{
+	if (space_left < NOTIFY_BUFFER_STATUS_THRESHOLD ||
+	    k_uptime_delta(&msg_notify_last_sent) >= mtrace_aging_timer) {
+		ipc_send_buffer_status_notify();
+		msg_notify_last_sent = k_uptime_get();
+
+		mtrace_pending = 0;
+
+	} else {
+		mtrace_pending += written;
+	}
+}
+
+static enum task_state mtrace_task(void *data)
+{
+	if (k_uptime_delta(&msg_notify_last_sent) >= mtrace_aging_timer)
+		mtrace_log_hook(0, 0);
+
+	return SOF_TASK_STATE_RESCHEDULE;
+}
+
+int ipc4_set_enable_logs(bool first_block,
+			 bool last_block,
+			 uint32_t data_offset_or_size,
+			 char *data)
+{
+	const struct log_backend *log_backend = log_backend_adsp_mtrace_get();
+	struct ipc4_log_state_info *log_state;
+	uint64_t delay;
+
+	if (!(first_block && last_block)) {
+		LOG_ERR("log_state data is expected to be sent as one chunk");
+		return -EINVAL;
+	}
+
+	if (data_offset_or_size < sizeof(struct ipc4_log_state_info)) {
+		LOG_ERR("log_state too small data size: %u", data_offset_or_size);
+		return -EINVAL;
+	}
+
+	dcache_invalidate_region((__sparse_force void __sparse_cache *)data, data_offset_or_size);
+
+	log_state = (struct ipc4_log_state_info *)data;
+
+	if (log_state->enable) {
+		adsp_mtrace_log_init(mtrace_log_hook);
+
+		log_backend_activate(log_backend, mtrace_log_hook);
+
+		mtrace_aging_timer = log_state->aging_timer_period;
+
+		schedule_task_init_ll(&mtrace_work,
```

Comment:
Any reason why this does not use Zephyr native scheduling API ? Is the intention to run this atm as a task in the LL domain so that we have guaranteed DSP time to get trace out (and since the data can fill the buffer quickly).

I suppose this can be addressed once the scheduling API is cleaned up.

Reply:
@lgirdwood I tried first with k_timer and then with k_work, but the problem I face is that the callback comes in the wrong context. I could set up a k_thread pinned to core 0, and run a k_work there, but then I'm basicly rewriting the boilerplate we have in zephyr_ll.c already. I don't particularly like this, but this is the only version I could get to reliably work for me.

Reply:
What's the requirement that it run on core 0?  Seems like for "idle work" kind of things, symmetric scheduling is generally a better fit anyway (i.e. it will never impact latency of the high priority audio work).

But that said: it would be perfectly legal to pin the system work queue thread to a specific core in application code, and if SOF wants its low priority stuff always on core 0, then that seems like it would work.

Reply:
@andyross This is actually just an SOF-side application design choice. All the IPC code is designed to run on single core, with no protection for multicore access. That single core just happens to be 0. And my empirical tests confirmed that basic tests start failing if I allow the IPC notifications to be sent from other cores. So this is more of a "If the mountain will not come to Mohammed, Mohammed will go to the mountain" case. Either I make the IPC stack multicore safe, or I accept the status quo and send my notifies from the IPC core. :)

But, but, I'll probably make another attempt at using a native method for this today. One big drawback with using the SOF ll-scheduler interface is that it'll get called every 1ms. This is very wasteful as the aging timer for logging is typically configured to 250ms and I can see it could be much slower than that.

Reply:
Ok, @lgirdwood @andyross now implemented with a native k_thread and k_sleep. Wasted a few hours when I had accidentally set the priority to that of the idle thread and my thread was not scheduled at all when other parts of SOF were not active, but once I figure out my mistake, this started to work. I used thread pinning to ensure the thread only gets scheduled on the same code as IPC stack is running on. Still not super pretty, but fits the current design and makes it explicitly why it is done so. The code could still be improved by having a k_timer and reprogramming that every time  a notification is sent based on buffer fill level, but not sure this is worth the extra code anymore.